src/arrow/schema.rs

```diff
@@ -0,0 +1,992 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Provides API for converting parquet schema to arrow schema and vice versa.
+//!
+//! The main interfaces for converting parquet schema to arrow schema  are
+//! `parquet_to_arrow_schema` and `parquet_to_arrow_schema_by_columns`.
+//!
+//! The interfaces for converting arrow schema to parquet schema is coming.
+
+use std::collections::HashSet;
+use std::rc::Rc;
+
+use schema::types::SchemaDescPtr;
+use schema::types::Type;
+use schema::types::TypePtr;
+use basic::LogicalType;
+use basic::Type as PhysicalType;
+use basic::Repetition;
+
+use arrow_lib::datatypes::Schema;
+use arrow_lib::datatypes::Field;
+use arrow_lib::datatypes::DataType;
+
+use errors::ParquetError::ArrowError;
+use errors::Result;
+
+/// Convert parquet schema to arrow schema.
+pub fn parquet_to_arrow_schema(parquet_schema: SchemaDescPtr) -> Result<Schema> {
+  parquet_to_arrow_schema_by_columns(parquet_schema.clone(),
+                                     0..parquet_schema.columns().len())
+}
+
+/// Convert parquet schema to arrow schema, only preserving some leaf columns.
+pub fn parquet_to_arrow_schema_by_columns<T>(
+  parquet_schema: SchemaDescPtr, column_indices: T) -> Result<Schema>
+  where T: IntoIterator<Item=usize> {
+  let mut base_nodes = Vec::new();
+  let mut base_nodes_set = HashSet::new();
+  let mut leaves = HashSet::new();
+  
+  for c in column_indices {
+    let column = parquet_schema.column(c).self_type() as *const Type;
+    let root = parquet_schema.get_column_root_ptr(c);
+    let root_raw_ptr = root.clone().as_ref() as *const Type;
+    
+    leaves.insert(column);
+    if !base_nodes_set.contains(&root_raw_ptr) {
+      base_nodes.push(root);
+      base_nodes_set.insert(root_raw_ptr);
+    }
+  }
+  
+  let leaves = Rc::new(leaves);
+  base_nodes.into_iter()
+    .map(|t| ParquetTypeConverter::new(t, leaves.clone()).to_field())
+    .collect::<Result<Vec<Option<Field>>>>()
+    .map(|result| result.into_iter()
+      .filter_map(|f| f)
+      .collect::<Vec<Field>>())
+    .map(|fields| Schema::new(fields))
+}
+
+/// This struct is used to group methods and data structures used to convert parquet
+/// schema together.
+struct ParquetTypeConverter {
+  schema: TypePtr,
+  leaves: Rc<HashSet<*const Type>>
```

Comment:
Why is `leaves` an `Rc<HashSet>`?

Reply:
I think the name `leaves` is quite misleading. It's not the leaves of schema, but the leave columns that needs to convert. I'll change that.